ts/ui/lazy/LazyHandler.ts

```diff
@@ -0,0 +1,483 @@
+ * @template D  The Document class
+ * @template B  The MathDocument class to extend
+ */
+export function LazyMathDocumentMixin<N, T, D,
+B extends MathDocumentConstructor<HTMLDocument<N, T, D>>>(
+  BaseDocument: B
+): MathDocumentConstructor<HTMLDocument<N, T, D>> & B {
+
+  return class BaseClass extends BaseDocument {
+
+    /**
+     * The Intersection Observer used to track the appearance of the expression markers
+     */
+    public lazyObserver: IntersectionObserver;
+
+    /**
+     * The mapping of markers to MathItems
+     */
+    public lazyList: LazyList<N, T, D>;
+
+    /**
+     * A promise to make sure our compiling/typesetting is sequential
+     */
+    protected lazyPromise: Promise<void> = Promise.resolve();
+
+    /**
+     * The function used to typeset a set of lazy MathItems.
+     *   (uses requestIdleCallback if available, or setTimeout otherwise)
+     */
+    protected lazyProcessSet: () => void;
+
+    /**
+     * True when a set of MathItems is queued for being processed
+     */
+    protected lazyIdle: boolean = false;
+
+    /**
+     * The set of items that have come into view
+     */
+    protected lazySet: LazySet = new Set();
+
+    /**
+     * Augment the MathItem class used for this MathDocument,
+     *   then create the intersection observer and lazy list,
+     *   and bind the lazyProcessSet function to this instance
+     *   so it can be used as a callback more easily.
+     *
+     * @override
+     * @constructor
+     */
+    constructor(...args: any[]) {
+      super(...args);
+      this.options.MathItem =
+        LazyMathItemMixin<N, T, D, Constructor<HTMLMathItem<N, T, D>>>(this.options.MathItem);
+      this.lazyObserver = new IntersectionObserver(this.lazyObserve.bind(this));
+      this.lazyList = new LazyList<N, T, D>();
+      const callback = this.lazyHandleSet.bind(this);
+      this.lazyProcessSet = (typeof window !== 'undefined' && window.requestIdleCallback ?
+                          () => window.requestIdleCallback(callback) :
+                          () => setTimeout(callback, 10));
+    }
+
+    /**
+     * The function used by the IntersectionObserver to monitor the markers coming into view.
+     * When one (or more) does, add it to or remove it from the set to be processed, and
+     *   if added to the set, queue an idle task, if one isn't already pending.
+     *
+     * @param {IntersectionObserverEntry[]} entries   The markers that have come into or out of view.
+     */
+    protected lazyObserve(entries: IntersectionObserverEntry[]) {
+      for (const entry of entries) {
+        const id = this.adaptor.getAttribute(entry.target as any as N, LAZYID);
+        const math = this.lazyList.get(id);
+        if (!math) return;
+        if (entry.isIntersecting) {
+          this.lazySet.add(id);
+          if (!this.lazyIdle) {
+            this.lazyIdle = true;
+            this.lazyProcessSet();
+          }
+        } else {
+          this.lazySet.delete(id);
```

Comment:
So if there is already a processor set that keeps running?

Reply:
Again, this is about removing an expression that you have scrolled past without stopping.  That is, if you start scrolling and reveal an expression marker, we add it into the set and queue an idle task.  But if you keep scrolling and the idle task hasn't run yet, the mark may move out of view, and we don't want to waste time typesetting it, so we remove it from the set.  When you stop scrolling and the idle task finally runs, it takes the expressions still in the set (those should be the ones still in view) and starts typesetting them after having created a new set to add expression to.  If one fo the expressions loads an extension and the idle task has to pause, you can add new expressions into the new list (or remove them from there), but can't affect the current idle-task list.  Those will be typeset even if they scroll out of view at this point.

It is possible that an expression is added to the set and then removed before the idle task runs, and it could be that the set is empty when it does.  Then the idle task does nothing, and new expressions are added to the new set to be processed by the next idle task.

Does that make sense?

Reply:
It does, and that's pretty much how I thought it worked. (And now have also practically experimented with.)
And I feel the gist of your last two comments should be part of the documentation for the `lazy` extension, to be kept for posterity.

Reply:
I have added the two paragraphs to the comments for the method.